### PR TITLE
Fix move's focus bugs

### PR DESCRIPTION
**Pull request by orestisf1993**
This will also be used for https://github.com/i3/i3/pull/2941 if accepted.

### CODE REVIEW

File: include/con.h [hunk 1]
```diff
@@ -115,6 +115,12 @@ Con *con_parent_with_orientation(Con *con, orientation_t orientation);
 Con *con_get_fullscreen_con(Con *con, fullscreen_mode_t fullscreen_mode);
 
 /**
+ * Returns the lowest container in the tree that has both a and b as descendants.
+ *
+ */
+Con *lowest_common_ancestor(Con *a, Con *b);
```
Comment (Airblader) [inline include/con.h]: Please name the function `con_*` as all the others.

File: src/con.c [hunk 2]
```diff
@@ -488,6 +488,25 @@ Con *con_get_fullscreen_con(Con *con, fullscreen_mode_t fullscreen_mode) {
...
+        Con *parent_b = b;
+        while (parent_b) {
+            if (parent_a == parent_b) {
+                return parent_a;
+            }
+            parent_b = parent_b->parent;
+        }
+        parent_a = parent_a->parent;
+    }
+    return NULL;
```
Comment (Airblader) [inline src/con.c]: This should never happen, right? So we should probably `assert(false)` here.

File: src/move.c [hunk 3]
```diff
@@ -24,6 +24,49 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
      * afterwards which might then close the con if it is empty. */
     Con *old_parent = con->parent;
 
+    Con *lca = lowest_common_ancestor(con, parent);
```
Comment (Airblader) [inline src/move.c]: We should drop a comment here explaining what the following lines are achieving.

File: src/move.c [hunk 4]
```diff
@@ -24,6 +24,49 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
...
+    Con *ancestor = con;
+    bool moves_focus_from_ancestor = true;
+    while (ancestor->parent != lca) {
```
Comment (Airblader) [inline src/move.c]: If `con` and `parent` were on different outputs (mot just workspaces) (let's put aside whether that makes sense for a second), then the LCA would be the root node. Which means `ancestor->parent == NULL`, no?

File: src/move.c [hunk 5]
```diff
@@ -24,6 +24,49 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
...
+        moves_focus_from_ancestor &= TAILQ_FIRST(&(ancestor->parent->focus_head)) == ancestor;
+        ancestor = ancestor->parent;
+    }
+    Con *target_ancestor = target;
+    while (target_ancestor->parent != lca) {
+        target_ancestor = target_ancestor->parent;
+    }
+
+    /* To determine if con is going to be placed before or after target in the parent's
+     * focus stack, we look at the stack order of the children of the least common
```
Comment (Airblader) [inline src/move.c]: s/least/lowest?

File: src/con.c (hunk 2 above)
Comment (orestisf1993) [reply src/con.c]: Ok. A hypothetical use could be if the caller wants to split the tree somewhere temporarily and then call the function and check for `NULL`. Will make the change.

File: src/move.c (hunk 4 above)
Comment (orestisf1993) [reply src/move.c]: `ancestor` and `target_ancestor` should always be children of `lca`, so no. Maybe I should give better names (ideas?) and write a comment as you said in another comment.

File: src/move.c [hunk 6]
```diff
@@ -24,6 +24,52 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
      * afterwards which might then close the con if it is empty. */
     Con *old_parent = con->parent;
 
+    /* We compare the focus order of the children of the least common ancestor. If con or
+     * it's ancestor is before target's ancestor then con should be placed before target
+     * in the focus stack. */
+    Con *lca = con_lowest_common_ancestor(con, parent);
+    Con *ancestor = con;
+    bool moves_focus_from_ancestor = true;
+    while (ancestor->parent != lca) {
```
Comment (orestisf1993) [inline src/move.c]: Here, if `con == lca` we will make i3 segfault. Should we assume that `con` can't be an ancestor of `target`? If yes, should I add an assertion?

File: src/move.c [hunk 7]
```diff
@@ -24,6 +24,52 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
...
+        moves_focus_from_ancestor &= TAILQ_FIRST(&(ancestor->parent->focus_head)) == ancestor;
+        ancestor = ancestor->parent;
+    }
+    Con *target_ancestor = target;
+    while (target_ancestor->parent != lca) {
```
Comment (Airblader) [inline src/move.c]: Perhaps extracting this logic into a `con_child_containing` function would improve readability and make this a bit more concise. We could also use it above by defining another function to turn it into something like

```
Con *ancestor = con_child_containing(lca, con);
bool moves_focus_from_ancestor = con_focused_inside(ancestor, con);
```

whereas the latter function checks that `con` is the focus head of `ancestor`, descending down the tree as necessary. This would mean that we need a few more iterations, but I think those cycles are well spent improving the readability of the change (which currently is my main concern here).

File: src/move.c [hunk 8]
```diff
@@ -24,6 +24,52 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
...
+        target_ancestor = target_ancestor->parent;
+    }
+
+    /* To determine if con is going to be placed before or after target in the parent's
+     * focus stack, we look at the stack order of the children of the lowest common
+     * ancestor. */
+    Con *current;
+    bool focus_before;
+    TAILQ_FOREACH(current, &(lca->focus_head), focused) {
```
Comment (Airblader) [inline src/move.c]: Perhaps this, too, would be worth refactoring into a `TAILQ_FIRST_OF` kind of function / macro used like

    TAILQ_FIRST_OF(&(lca->focus_head), ancestor, target_ancestor, focused)

and returning whichever of the two comes first (which is what this does). I know that @stapelberg generally prefers only extracting something used more than once, but I think the abstraction would help simplifying the code and improving readability. What do you think?

File: src/move.c [hunk 9]
```diff
@@ -24,6 +24,52 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
...
+     * before the old ancestor in the focus stack. Example:
+     * Consider the layout [ H [ V1 [ A B ] V2 [ C ] ] ] where A is focused. We move to
+     * a second workspace and from there we move A to the right and switch back to the
+     * original workspace. Without the change focus would move to B instead of staying
+     * with A. */
+    if (moves_focus_from_ancestor && focus_before) {
+        Con *place = TAILQ_PREV(ancestor, focus_head, focused);
+        TAILQ_REMOVE(&(lca->focus_head), target_ancestor, focused);
+        if (place) {
+            TAILQ_INSERT_AFTER(&(lca->focus_head), place, target_ancestor, focused);
```
Comment (Airblader) [inline src/move.c]: I think we have this kind of logic more than once in the code already. Perhaps a `TAILQ_INSERT_AFTER_OR_HEAD` would be good to extract. To not blow up the PR we could keep this as an option for a followup PR to explicitly do this kind of refactoring, though. Similarly, we might be able to use a `TAILQ_INSERT_BEFORE_OR_TAIL` in other places.

File: src/move.c [hunk 10]
```diff
@@ -24,6 +24,52 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
      * afterwards which might then close the con if it is empty. */
     Con *old_parent = con->parent;
 
+    /* We compare the focus order of the children of the least common ancestor. If con or
+     * it's ancestor is before target's ancestor then con should be placed before target
```
[the remark this answers is not in the document]
Comment (orestisf1993) [reply src/move.c]: I think I'll bind "it's" to nagbar or something, I make this mistake 30 times per day.

[the remark this answers is not in the document]
Comment (Airblader) [reply src/move.c]: Probably not a good idea – it's and its both exist, it just depends on what it stands for. :-)

[the remark this answers is not in the document]
Comment (orestisf1993) [reply src/move.c]: Just to be aware of what I just typed

File: src/move.c (hunk 7 above)
Comment (orestisf1993) [reply src/move.c]: Is `con_focused_inside` a good name? It might look like it is meant to check for `focused` inside `ancestor`.

Comment (orestisf1993) [reply src/move.c]: > into a `con_child_containing` function

maybe `con_ancestor_with_parent(Con* parent, Con* con)`?

File: src/move.c (hunk 8 above)
Comment (orestisf1993) [reply src/move.c]: Is there a good way to have a macro that declares something (`current` for `TAILQ_FOREACH`) and "return" something? I think it should be a function.

File: src/move.c (hunk 7 above)
Comment (Airblader) [reply src/move.c]: I'm open to other names, but generally speaking "child" is both a well known graph theoretical term and an established term in our code base. But feel free to propose names as we make the changes and then we can still revisit them.

File: src/move.c (hunk 8 above)
Comment (Airblader) [reply src/move.c]: `GREP_FIRST` would be an example.

Comment (orestisf1993) [reply src/move.c]: IMO an inline function will be cleaner here because the `GREP_FIRST` way requires an extra argument.

Comment (Airblader) [reply src/move.c]: Fair enough. In that case we could also just make it specific and turn it into something like `con_first_focused_of(Con *parent, Con *a, Con *b)` or something like that.

File: src/move.c (hunk 7 above)
Comment (orestisf1993) [reply src/move.c]: > Is `con_focused_inside` a good name? It might look like it is meant to check for `focused` inside `ancestor`.

Especially since we already have `con_inside_focused(Con *con)` which checks for the `focused` container.

File: src/move.c (hunk 8 above)
Comment (orestisf1993) [reply src/move.c]: Changed my mind again, a macro can use `field`, generalizing it for other uses.

It should be enough to modify `GREP_FIRST` to have a `field` argument.

Comment (orestisf1993) [reply src/move.c]: What I have right now is:

```
#define TAILQ_FIRST_OF(head, destelm, elm1, elm2, field) \
    TAILQ_FOREACH(destelm, head, field) {                \
        if (destelm == elm1 || destelm == elm2) {        \
            break;                                       \
        }                                                \
    }
```

Using it like this:
```
Con *current;
TAILQ_FIRST_OF(&(lca->focus_head), current, ancestor, target_ancestor, focused);
bool focus_before = current == ancestor;
```

Break condition could be generalized to `condition`. Otherwise, I think it's possible to have a variable number of `elmx` arguments.

Comment (Airblader) [reply src/move.c]: We don't need to over-generalize the function for now. Once we need it, we can refactor it some more.

File: src/move.c (hunk 7 above)
Comment (Airblader) [reply src/move.c]: Fair enough. Just propose a name you think fits and we can see what it looks like. :-) Names are easy to change.

**Comment by orestisf1993** [comment]
Added the functions.

File: testcases/t/539-move-focus.t [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+$D = open_window;
+is($x->input_focus, $D->id, 'Window D focused');
+
+$ws2 = fresh_workspace;
+cmd '[id=' . $A->id . '] move right';
+cmd "workspace $ws1";
+sync_with_i3;
+
+is($x->input_focus, $D->id, 'Window D still focused');
+kill_and_confirm_focus($C->id, 'C focused net');
```
Comment (Airblader) [inline testcases/t/539-move-focus.t]: net?

File: testcases/t/539-move-focus.t [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+kill_and_confirm_focus($B->id, 'C focused net');
```
Comment (Airblader) [inline testcases/t/539-move-focus.t]: Should be B here and A below, right?

File: testcases/t/113-urgent.t [hunk 13]
```diff
@@ -333,6 +333,35 @@ for ($type = 1; $type <= 2; $type++) {
     is($target_ws->{urgent}, 1, 'Target workspace is now marked urgent');
 
 ##############################################################################
+# Test that moving an unfocused container doesn't reset it's urgency hint.
+# See issue:
```
Comment (Airblader) [inline testcases/t/113-urgent.t]: assert(issue) :-)

File: testcases/t/113-urgent.t [hunk 14]
```diff
@@ -333,6 +333,35 @@ for ($type = 1; $type <= 2; $type++) {
     is($target_ws->{urgent}, 1, 'Target workspace is now marked urgent');
 
 ##############################################################################
+# Test that moving an unfocused container doesn't reset it's urgency hint.
```
Comment (Airblader) [inline testcases/t/113-urgent.t]: its :-)

File: src/con.c [hunk 15]
```diff
@@ -817,6 +836,35 @@ int con_num_visible_children(Con *con) {
...
+/*
+ * Returns true if the given container is the most focused descendant of ancestor,
+ * recursively.
+ *
+ */
+bool con_is_most_focused_descendant(Con *con, Con *ancestor) {
+    Con *current = con;
+    bool result = true;
+    while (result && current != ancestor) {
+        result &= TAILQ_FIRST(&(current->parent->focus_head)) == current;
```
Comment (Airblader) [inline src/con.c]: I think we can rewrite this to

```
while (…) {
    if (!…) {
        return false;
    }

    …
}

return true;
```

This will break out of the loop as soon as we know the result will be false, rather than always going through the entire hierarchy.

File: src/con.c [hunk 16]
```diff
@@ -817,6 +836,35 @@ int con_num_visible_children(Con *con) {
...
+Con *con_child_of_ancestor(Con *ancestor, Con *con) {
+    Con *result = con;
+    while (result && result->parent != ancestor) {
+        result = result->parent;
+    }
+    return result;
+}
+
+/*
+ * Returns true if the given container is the most focused descendant of ancestor,
```
Comment (Airblader) [inline src/con.c]: It doesn't seem intuitive to me what a »most focused« container is? Can we not just do without the »most«, i.e., `con_is_focused_descendant`?

File: src/con.c [hunk 17]
```diff
@@ -817,6 +836,35 @@ int con_num_visible_children(Con *con) {
 }
 
 /*
+ * Given an ancestor of a container and the container return the ancestor container of
+ * con that is a child of the given ancestor.
+ *
+ */
+Con *con_child_of_ancestor(Con *ancestor, Con *con) {
```
Comment (Airblader) [inline src/con.c]: I'm still not too happy with this name. What would you think about `con_child_containing`? A child is by definition a direct descendant, so the terminology should be clear.

File: include/queue.h [hunk 18]
```diff
@@ -354,6 +354,13 @@
          (var) != TAILQ_END(head);                        \
          (var) = TAILQ_PREV(var, headname, field))
 
+#define TAILQ_FIRST_OF(destelm, head, elm1, elm2, field) \
```
Comment (Airblader) [inline include/queue.h]: Let's maybe add a small explanation along the lines of

> Returns elm1, elm2 or NULL in destelm when going through the tail queue, depending on which one comes first.

File: src/con.c (hunk 15 above)
Comment (orestisf1993) [reply src/con.c]: I do break if result is false but I can use an if if you prefer it.

File: testcases/t/539-move-focus.t (hunk 12 above)
Comment (orestisf1993) [reply testcases/t/539-move-focus.t]: I don't know why these ended up here. I corrected them somewhere but probably checked out something else somewhere?

File: src/con.c (hunk 17 above)
Comment (orestisf1993) [reply src/con.c]: It's the child of the ancestor of con. Not too happy with the name either.

Can you explain the `con_child_containing` name?

File: testcases/t/113-urgent.t (hunk 14 above)
Comment (orestisf1993) [reply testcases/t/113-urgent.t]: 😭

File: include/queue.h (hunk 18 above)
Comment (orestisf1993) [reply include/queue.h]: Ok. The rest didn't have any explanation, that's why I didn't add any.

File: src/con.c (hunk 15 above)
Comment (Airblader) [reply src/con.c]: Oops :-) But yeah, I do think the other way is easier to read and avoids the result variable, so let's change it.

File: src/con.c (hunk 17 above)
Comment (Airblader) [reply src/con.c]: It's the container's child (== direct descendant) which contains the given container.

Comment (orestisf1993) [reply src/con.c]: Ok I'll use the name then.

Just one quick question if `A->parent->parent->...->parent == B` do we say that B **contains** A?

Comment (Airblader) [reply src/con.c]: OK, so the question is whether »containing« is defined as »is ancestor of« or »is parent of«. I don't think we have a clear definition of this right now; in some parts we make this clear by appending `_recursively`, which I wouldn't mind doing here if it helps clearing things up.

**Comment by orestisf1993** [comment]
@Airblader I've been wondering if it gets harder to track changes to a bigger PR after lots of reviews. Would you prefer it if I pushed `--fixup=` and `--squash=` type of commits and then rebased before the final merge?

**Comment by orestisf1993** [comment]
Up

File: src/move.c (hunk 9 above)
Comment (orestisf1993) [reply src/move.c]: Implemented this, tell me if you want to push it here after all. I also think that a macro
`#define TAILQ_INSERT_POSITION(head, listelm, elm, position, field)`
 that does everything will be handy:

- Compares `position` to `BEFORE` and `AFTER` (`position_t`)
- Uses `TAILQ_INSERT_AFTER_OR_HEAD` or `TAILQ_INSERT_BEFORE_OR_TAIL` accordingly.

Comment (orestisf1993) [reply src/move.c]: hmm, the code of this PR uses BEFORE-HEAD and AFTER-TAIL.

Comment (orestisf1993) [reply src/move.c]: I put the changes here, tell me if you want to push them in this PR: https://github.com/orestisf1993/i3/tree/move-con_focus-tmp

**Comment by Airblader** [comment]
LGTM now, but I'd like @stapelberg to at least glance over it as well.

File: src/move.c [hunk 19]
```diff
@@ -24,6 +24,42 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
...
+    Con *con_ancestor = con_child_containing_recursively(lca, con);
+    Con *target_ancestor = con_child_containing_recursively(lca, target);
+    bool moves_focus_from_ancestor = con_is_focused_descendant(con, con_ancestor);
+
+    /* To determine if con is going to be placed before or after target in the parent's
+     * focus stack, we look at the stack order of the children of the lowest common
+     * ancestor. */
+    Con *current;
+    TAILQ_FIRST_OF(current, &(lca->focus_head), con_ancestor, target_ancestor, focused);
+    bool focus_before = current == con_ancestor;
```
Comment (stapelberg) [inline src/move.c]: nit: `bool focus_before = (current == con_ancestor);` for consistency

File: src/move.c [hunk 20]
```diff
@@ -24,6 +24,42 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
      * afterwards which might then close the con if it is empty. */
     Con *old_parent = con->parent;
 
+    /* We compare the focus order of the children of the least common ancestor. If con or
```
Comment (stapelberg) [inline src/move.c]: nit: s/least/lowest/

File: src/con.c [hunk 21]
```diff
@@ -817,6 +836,35 @@ int con_num_visible_children(Con *con) {
 }
 
 /*
+ * Given an ancestor of a container and the container return the child of ancestor that
+ * recursively contains con.
```
Comment (stapelberg) [inline src/con.c]: I find this description hard to follow. How about the following instead?

> Returns the direct child of ancestor that contains con.

File: src/con.c [hunk 22]
```diff
@@ -817,6 +836,35 @@ int con_num_visible_children(Con *con) {
...
+ *
+ */
+Con *con_child_containing_recursively(Con *ancestor, Con *con) {
+    Con *child = con;
+    while (child && child->parent != ancestor) {
+        child = child->parent;
+    }
+    return child;
```
Comment (stapelberg) [inline src/con.c]: Add `assert(child->parent);` to catch issues early (like in `con_lowest_common_ancestor`).

File: src/move.c [hunk 23]
```diff
@@ -24,6 +24,42 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
...
+
+    Con *con_ancestor = con_child_containing_recursively(lca, con);
+    Con *target_ancestor = con_child_containing_recursively(lca, target);
+    bool moves_focus_from_ancestor = con_is_focused_descendant(con, con_ancestor);
+
+    /* To determine if con is going to be placed before or after target in the parent's
+     * focus stack, we look at the stack order of the children of the lowest common
+     * ancestor. */
+    Con *current;
+    TAILQ_FIRST_OF(current, &(lca->focus_head), con_ancestor, target_ancestor, focused);
```
Comment (stapelberg) [inline src/move.c]: I think this should be expressed inline to achieve more clarity:
```c
Con *current, *first = NULL;
TAILQ_FOREACH(current, &(lca->focus_head), focused) {
  if (current == con_ancestor || current == target_ancestor) {
    first = current;
    break;
  }
}
```
That way, readers of the code don’t need to look up what `TAILQ_FIRST_OF` does (i.e. this approach reduces mental complexity of an already very complicated function).

File: include/con.h [hunk 24]
```diff
@@ -115,6 +115,12 @@ Con *con_parent_with_orientation(Con *con, orientation_t orientation);
 Con *con_get_fullscreen_con(Con *con, fullscreen_mode_t fullscreen_mode);
 
 /**
+ * Returns the lowest container in the tree that has both a and b as descendants.
+ *
+ */
+Con *con_lowest_common_ancestor(Con *a, Con *b);
```
Comment (stapelberg) [inline include/con.h]: Regarding all the new functions in con.c, I think it’d make more sense to place them in move.c as `static` functions until we need them elsewhere in the code base.

File: src/move.c [hunk 25]
```diff
@@ -46,12 +82,24 @@ static void insert_con_into(Con *con, Con *target, position_t position) {
 
     con->parent = parent;
 
+    if (parent == lca) {
```
Comment (stapelberg) [inline src/move.c]: Could you add examples to each of the branches here please? Ideally using unicode drawing characters, like in testcases/t/510-focus-across-output.t

File: src/move.c (hunk 23 above)
Comment (Airblader) [reply src/move.c]: That's what we had at first; I actually see it entirely from the opposite side. Inlining just introduces more branches and more complexity. With a well-named function we'd be abstracting that logic and only if the reader needs to know specific details of the function (or macro), they'd be looking up the definition. This is already a really long function, so to me breaking it up into pieces makes a lot of sense.

Comment (stapelberg) [reply src/move.c]: Sure, breaking up functions into self-contained pieces is a good idea in general. Maybe we just haven’t identified the correct pieces yet.

If we absolutely want to stick to having this bit of code abstracted away, let’s at least scope the macro to move.c, not to the entire code base.

Comment (orestisf1993) [reply src/move.c]: IMO moving the scope to move.c is not a great idea:

1. The macro is not exclusive to move.c semantically. It could be relevant to any place that uses TAILQs.
Not sure if I phrased this correctly.
2. Breaking `TAILQ_` macros in many files makes it harder to be aware of all of them and could lead to code repetition.

File: src/move.c (hunk 25 above)
Comment (orestisf1993) [reply src/move.c]: Ok. One alternative would be to draw tree structures which could make it easier to understand the code for focus comparison between ancestors. A drawing of a workspace is closer to real life though.

btw, are there any tools that can help me with this?

**Comment by orestisf1993** [comment]
:| move.c's `attach_to_workspace` suffers from the same issues.

**Comment by Airblader** [comment]
What's the status on this one? Anything open from your point of view or should we get into the review process to merge this?

**Comment by orestisf1993** [comment]
Yes you can continue the review process. I have left some items open from @stapelberg's review because I am not so sure about them.

As I pointed out `attach_to_workspace` has issues as well but they don't need to be fixed here.

**Comment by Airblader** [comment]
I'm sorry it's been almost two weeks again, I'm just really lacking time right now. Perhaps @stapelberg can review it since it's his comments that are open anyway. :-)

File: src/move.c (hunk 25 above)
Comment (stapelberg) [reply src/move.c]: I’m not aware of any i3-specific tools. There are tools which allow you to draw with Unicode drawing characters, but I haven’t tested any of them.

File: src/move.c (hunk 23 above)
Comment (stapelberg) [reply src/move.c]: That’s fair. Let’s remove the macro then, until we use it in multiple places.

**Comment by orestisf1993** [comment]
Up. I've added a new branch regarding `focus_before`. I found the problem because I couldn't find a case where `parent == lca` and `focus_before == 1`.

I think I changed everything @stapelberg mentioned.
